lib/msal-react/src/hooks/useMsalAuthentication.ts

```diff
@@ -21,46 +21,56 @@ export function useMsalAuthentication(
     const [[result, error], setResponse] = useState<[AuthenticationResult|null, AuthError|null]>([null, null]);
     const [hasBeenCalled, setHasBeenCalled] = useState<boolean>(false);
 
-    const login = useCallback(async (loginType: InteractionType, request?: PopupRequest|RedirectRequest|SsoSilentRequest): Promise<AuthenticationResult|null> => {
+    const login = useCallback(async (callbackInteractionType?: InteractionType, callbackRequest?: PopupRequest|RedirectRequest|SsoSilentRequest): Promise<AuthenticationResult|null> => {
+        const loginType = callbackInteractionType || interactionType;
+        const loginRequest = callbackRequest || authenticationRequest || {};
         switch (loginType) {
             case InteractionType.Popup:
-                return instance.loginPopup(request as PopupRequest);
+                return instance.loginPopup(loginRequest as PopupRequest);
             case InteractionType.Redirect:
                 // This promise is not expected to resolve due to full frame redirect
-                return instance.loginRedirect(request as RedirectRequest).then(null);
+                return instance.loginRedirect(loginRequest as RedirectRequest).then(null);
             case InteractionType.Silent:
-                return instance.ssoSilent(request as SsoSilentRequest);
+                return instance.ssoSilent(loginRequest as SsoSilentRequest);
             default:
                 throw "Invalid interaction type provided.";
         }
+    }, [instance, interactionType, authenticationRequest]);
+
+    useEffect(() => {
+        const callbackId = instance.addEventCallback((message: EventMessage) => {
+            switch(message.eventType) {
+                case EventType.LOGIN_SUCCESS:
+                case EventType.SSO_SILENT_SUCCESS:
+                    message.payload && setResponse([message.payload as AuthenticationResult, null]);
+                    break;
+                case EventType.LOGIN_FAILURE:
+                case EventType.SSO_SILENT_FAILURE:
+                    message.error && setResponse([null, message.error as AuthError]);
+                    break;
+            }
+        });
+
+        return () => {
+            callbackId && instance.removeEventCallback(callbackId);
+        };
     }, [instance]);
 
     useEffect(() => {
         if (!hasBeenCalled && !isAuthenticated && inProgress === InteractionStatus.None) {
             // Ensure login is only called one time from within this hook, any subsequent login attempts should use the callback returned
             setHasBeenCalled(true);
-            login(interactionType, authenticationRequest)
-                .then(result => {
-                    isMounted.current && setResponse([result, null]);
-                })
-                .catch(error => {
-                    isMounted.current && setResponse([null, error]);
-                });
+            login().catch(() => {
```

Comment:
Why do we still need the catch?

Reply:
Otherwise errors will throw and potentially break application flow since we are surfacing the errors in a different way